### PR TITLE
feat: refine instruction generation and add reflection option to MIPR…

### PR DESCRIPTION
**Pull request by shamuiscoding**
…O.optimize

### CODE REVIEW

File: ape/ape_prompts/extract-keys.prompt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+response_format:
+    type: json_schema
+    json_schema:
+        name: KeysResponse
+        schema:
+            type: object
+            properties:
+                answer:
+                    type: string
+                reasoning:
```
Comment (aschung01) [inline ape/ape_prompts/extract-keys.prompt:17]: Shouldn't the reasoning come before the answer? @shamuiscoding

File: ape/ape_prompts/extract-keys.prompt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+Find the keys corresponding to the answer and reasoning in the response.
+
+If there are no keys corresponding to the reasoning, then set reasoning as None.
+
+<EXAMPLE>
+<input>
+{
+    "answer": "123",
+    "reasoning": "Example Reasoning"
```
Comment (aschung01) [inline ape/ape_prompts/extract-keys.prompt]: here as well

File: ape/ape_prompts/gen-instructions.prompt [hunk 3]
```diff
@@ -47,7 +47,7 @@ If you would like to use few shot examples, use the reserved variable {{_FEWSHOT
 Use the information below to learn about a task that we are trying to solve using calls to an LM, then write a new prompt that will be used to prompt a Language Model to better solve the task.
 
 Task for LLM:
-{prompt_desc}
```
Comment (aschung01) [inline ape/ape_prompts/gen-instructions.prompt]: this is meant to be `prompt_desc` to describe the task of the prompt. task_description that comes below as a suggestion is to provide a tip (especially for human expertise)

File: ape/optimizer/mipro/mipro_proposer.py [hunk 4]
```diff
@@ -16,7 +16,6 @@ async def generate_candidates(
         self,
         trainset: Dataset,
         task_description: Optional[str] = None,
-        prompt_desc: Optional[str] = None,
```
Comment (aschung01) [inline ape/optimizer/mipro/mipro_proposer.py:19]: this shouldn't be removed as well

**Review by aschung01 — CHANGES_REQUESTED** [review]

File: ape/ape_prompts/extract-keys.prompt (hunk 2 above)
Comment (shamuiscoding) [reply ape/ape_prompts/extract-keys.prompt]: doesn't really matter as this is only for extracting the keys. but will change to keep consistency

File: ape/ape_prompts/gen-instructions.prompt (hunk 3 above)
Comment (shamuiscoding) [reply ape/ape_prompts/gen-instructions.prompt]: right now, there is no way of adding prompt_desc into optimize function, and that is why i removed it. 
i'll add prompt_desc as a param to mipro.optimize and revive the variable

File: ape/optimizer/mipro/mipro_proposer.py (hunk 4 above)
Comment (aschung01) [reply ape/optimizer/mipro/mipro_proposer.py:19]: @shamuiscoding Did you add this?

**Review by aschung01 — CHANGES_REQUESTED** [review]

File: ape/ape_prompts/reflexion.prompt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+model: claude-3-5-sonnet-20240620
+description: Create a new prompt based base prompt and reflection
+inputs: 
+    question: The question to be answered
+    llm_output: The output of the LLM
+    llm_reasoning: The reasoning of the LLM
+    ground_truth_reasoning: The reasoning of the ground truth
+    ground_truth_answer: The answer of the ground truth
+outputs:
+    prompt: A new prompt based on the base prompt and reflection
```
Comment (engineerA314) [inline ape/ape_prompts/reflexion.prompt:11]: it should be `question` and `insight`?

and `response_format  = {type:json_object}`is needed

File: ape/optimizer/mipro/mipro.py [hunk 6]
```diff
@@ -323,5 +389,61 @@ def objective(trial: optuna.Trial) -> float:
...
+                keys = await extract_keys(
+                    response=failed_samples[0]
+                )
+                
+                answer_key = keys.get("answer", None)
+                reasoning_key = keys.get("reasoning", None)
+                
+                if answer_key is None or reasoning_key is None:
+                    logger.warning("No answer or reasoning key found in the response. Skipping reflexion.")
+                    return best_prompt
```
Comment (engineerA314) [inline ape/optimizer/mipro/mipro.py:419]: reflexion is only works for data that have `answer` as output key?

File: ape/optimizer/mipro/mipro.py [hunk 7]
```diff
@@ -323,5 +389,61 @@ def objective(trial: optuna.Trial) -> float:
...
+                    "Here are the questions and insights on how to solve them.\n"
+                    "ONLY IF a question asking for a similar problem is asked, "
+                    "learn from the insights and solve the problem correctly.\n"
+                    "<START OF INSIGHT>\n"
+                    f"{''.join(reflections)}"
+                    "\n<END OF INSIGHT>\n"
+                    "Remember to apply these insights only when encountering similar problem types. "
+                    "Do not use them if the question is not related to the examples provided.\n\n"
+                )
+                best_prompt.messages[0]["content"] += reflexion_message
```
Comment (engineerA314) [inline ape/optimizer/mipro/mipro.py:440]: problem, question, answer.
I think it should be generalized.

**Pull request closed without merging**